src/app/credExplorer/weights/weights.js

```diff
@@ -0,0 +1,74 @@
+// @flow
+
+import {NodeAddress, EdgeAddress} from "../../../core/graph";
+import type {NodeType, EdgeType} from "../../adapters/pluginAdapter";
+import type {StaticPluginAdapter} from "../../adapters/pluginAdapter";
+import type {StaticAdapterSet} from "../../adapters/adapterSet";
+
+export type WeightedNodeType = {|+type: NodeType, +weight: number|};
+
+export type WeightedEdgeType = {|
+  +type: EdgeType,
+  +forwardWeight: number,
+  +backwardWeight: number,
+|};
+
+export type WeightedTypes = {|
+  +nodes: $ReadOnlyArray<WeightedNodeType>,
+  +edges: $ReadOnlyArray<WeightedEdgeType>,
+|};
+
+export function defaultWeightedNodeType(type: NodeType): WeightedNodeType {
+  return {
+    type,
+    weight: type.defaultWeight,
+  };
+}
+
+export function defaultWeightedEdgeType(type: EdgeType): WeightedEdgeType {
+  return {
+    type,
+    forwardWeight: type.defaultForwardWeight,
+    backwardWeight: type.defaultBackwardWeight,
+  };
+}
+
+export function defaultWeightsForAdapter(
+  adapter: StaticPluginAdapter
+): WeightedTypes {
+  return {
+    nodes: adapter.nodeTypes().map(defaultWeightedNodeType),
+    edges: adapter.edgeTypes().map(defaultWeightedEdgeType),
+  };
+}
+
+export function combineWeights(
+  ws: $ReadOnlyArray<WeightedTypes>
+): WeightedTypes {
+  const seenPrefixes = new Set();
```

Comment:
Note that if you didn’t care about saying specifically which prefix was duplicated, then this could just be

```js
if (new Set(nodes.map((x) => x.type.prefix)).size !== nodes.length) {
  throw new Error("Duplicate prefix among: " + JSON.stringify(nodes));
}
```

But this is fine with me, too.

Reply:
Ack.